crates/ruff_linter/src/rules/ruff/rules/post_init_default.rs

```diff
@@ -0,0 +1,90 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, ParameterWithDefault};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for `__post_init__` dataclass methods with argument defaults.
+///
+/// ## Why is this bad?
+/// Variables that are only used during initialization should be instantiated
+/// as an init-only pseudo-field using `dataclasses.InitVar`. According to the
+/// [documentation],
+///
+/// > Init-only fields are added as parameters to the generated `__init__()`
+/// > method, and are passed to the optional `__post_init__()` method. They are
+/// > not otherwise used by dataclasses.
+///
+/// Default values for `__post_init__` arguments that exist as init-only field
+/// as well will be overridden by the `dataclasses.InitVar` value.
+///
+/// ## Example
+/// ```python
+/// from dataclasses import InitVar, dataclass
+///
+///
+/// @dataclass
+/// class Foo:
+///     bar: InitVar[int] = 0
+///
+///     def __post_init__(self, bar: int = 1, baz: int = 2) -> None:
+///         print(bar, baz)
+///
+///
+/// foo = Foo()  # Prints '0 2'.
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from dataclasses import InitVar, dataclass
+///
+///
+/// @dataclass
+/// class Foo:
+///     bar: InitVar[int] = 1
+///     baz: InitVar[int] = 2
+///
+///     def __post_init__(self, bar, baz) -> None:
+///         print(bar, baz)
+///
+///
+/// foo = Foo()  # Prints '1 2'.
+/// ```
+///
+/// ## References
+/// - [Python documentation: Post-init processing](https://docs.python.org/3/library/dataclasses.html#post-init-processing)
+/// - [Python documentation: Init-only variables](https://docs.python.org/3/library/dataclasses.html#init-only-variables)
+///
+/// [documentation]: https://docs.python.org/3/library/dataclasses.html#init-only-variables
+#[violation]
+pub struct PostInitDefault;
+
+impl Violation for PostInitDefault {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`__post_init__` method has argument defaults, use `dataclasses.InitVar` instead")
+    }
+}
+
+/// RUF033
+pub(crate) fn post_init_default(checker: &mut Checker, function_def: &ast::StmtFunctionDef) {
+    if &function_def.name != "__post_init__" {
+        return;
+    }
+
+    for ParameterWithDefault {
+        parameter: _,
+        default,
+        range: _,
+    } in function_def.parameters.iter_non_variadic_params()
+    {
+        let Some(default) = default else {
+            continue;
+        };
+        checker
+            .diagnostics
+            .push(Diagnostic::new(PostInitDefault, default.range()));
+    }
+}
```

Comment:
Before emitting the diagnostic, could we also check:
1. That we're inside a class scope (we shouldn't emit for `__post_init__` functions in the global scope)
2. That the class is decorated with `@dataclasses.dataclass` (since `__post_init__` has no special meaning for other classes)

Reply:
Done in https://github.com/astral-sh/ruff/pull/13192/commits/6ca05468a85d7a9266d986205b2ccb57ae3f3301